File: 150_Evaluate Reverse Polish Notation.cpp
/*
150. Evaluate Reverse Polish Notation
Evaluate arithmetic expression given in Reverse Polish Notation (postfix) using a stack.
*/
#include <bits/stdc++.h>
using namespace std;

int evalRPN(vector<string>& tokens){
    stack<long long> st;
    for(auto &t: tokens){
        if(t=="+"||t=="-"||t=="*"||t=="/"){
            long long b = st.top(); st.pop();
            long long a = st.top(); st.pop();
            if(t=="+") st.push(a+b);
            else if(t=="-") st.push(a-b);
            else if(t=="*") st.push(a*b);
            else { st.push(a / b); }
        } else {
            st.push(stoll(t));
        }
    }
    return (int)st.top();
}

int main(){
    vector<string> t1 = {"2","1","+","3","*"};
    cout << "evalRPN([2,1,+,3,*]) = " << evalRPN(t1) << "\n"; // (2+1)*3 = 9

    vector<string> t2 = {"4","13","5","/","+"};
    cout << "evalRPN([4,13,5,/, +]) = " << evalRPN(t2) << "\n"; // 6

    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: 150_Evaluate Reverse Polish Notation.py
"""
150. Evaluate Reverse Polish Notation
Use a stack to evaluate tokens in postfix order.
"""
from typing import List

def evalRPN(tokens: List[str]) -> int:
    st = []
    for t in tokens:
        if t in ('+','-','*','/'):
            b = int(st.pop()); a = int(st.pop())
            if t == '+': st.append(a+b)
            elif t == '-': st.append(a-b)
            elif t == '*': st.append(a*b)
            else: st.append(int(a / b))
        else:
            st.append(int(t))
    return st[-1]

if __name__ == '__main__':
    print("evalRPN(['2','1','+','3','*']) =", evalRPN(['2','1','+','3','*']))
    print("evalRPN(['4','13','5','/','+']) =", evalRPN(['4','13','5','/','+']))

# Time Complexity: O(n)
# Space Complexity: O(n)
